### PR TITLE
Add `--checkpoint-on-crash`

### LINKED ISSUE #1468: Save diagnostics at the timestep before crashing
# Description

It would be useful for debugging to output the interpolated diagnostics (and/or dump un-interpolated state/aux variables) for the timestep before the model crashes. We have one crude version that runs on 1 rank in [this branch](https://github.com/CliMA/ClimateMachine.jl/tree/ln/save_crashed_tstep_1rank) (the functionality for ntasks=1 should be obtainable by just making the modifications in `src/Driver/Driver.jl` and `src/Diagnostics/atmos_gcm_default.jl` ) , but this takes a while. 

# Additional context

Add any other reasons why this feature will help with the goals of CLIMA.

# For CLIMA Developers

- [ ] This feature can be added (if it cannot be, explain why in a comment below -- lack of technical expertise, not relevant to the scope of this project, too ambitious)
- [ ] There is a timeline for when this feature can be implemented
- [ ] The feature has been (or will be) implemented (Please link the PR)

### LINKED ISSUE #1510: Heartbeat with synchronised error handling
# Description

We want to be able to dump the full state when an error occurs inside the DG kernels. The challenge is determining when this occurs and flagging this on all ranks. A simple idea is to do something like
```julia
try
   # do stuff
   err = MPI.Allreduce(false, max, comm)
catch e
   err = MPI.Allreduce(true, max, comm)
finally
   if err
      # write state to disk
      error("now quit")
   end
end
```

The problem with this approach is that if `# do stuff` involved communication, we might end up in a state where ranks without errors are stuck waiting on communication from the errored ranks.

Since our errors will mostly occur inside events, one idea is to build this into an event wrapper
```julia
struct HeartbeatEvent <: Event
   ev::Event
   callback
end
function wait(h::HeartbeatEvent)
   try
     wait(h.ev)
     err = MPI.Allreduce(false, max, comm)
   catch e
     err = MPI.Allreduce(true, max, comm)
   finally
     if err
      h.callback()
      error("now quit")
     end
   end
end
```
cc: @kpamnany @vchuravy 

# For CLIMA Developers

- [ ] This feature can be added (if it cannot be, explain why in a comment below -- lack of technical expertise, not relevant to the scope of this project, too ambitious)
- [ ] There is a timeline for when this feature can be implemented
- [ ] The feature has been (or will be) implemented (Please link the PR)

### PR DESCRIPTION
**Pull request by kpamnany**
### Description

If specified, `--checkpoint-on-crash` adds exception checking to the `wait` call in `MPIStateArrays` which is where most such crashes will be detected. An `Allreduce` call is added here which notifies all MPI ranks of the crash; a checkpoint is then written.

I've tested this using the baroclinic wave experiment (with the [filter](https://github.com/CliMA/ClimateMachine.jl/blob/master/experiments/TestCase/baroclinic_wave.jl#L288) removed to force a crash) on CPU and GPU with 1, 2 and 32 MPI ranks and it seems to work fine.

Closes #1468 and #1510.

- [X] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [X] Unit tests are included OR N/A.
- [X] Code is exercised in an integration test OR N/A.
- [X] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by LenkaNovak** [comment]
Works like a charm! 🏆

**Review by LenkaNovak — APPROVED** [review]

File: src/Arrays/MPIStateArrays.jl [hunk 1]
```diff
@@ -888,4 +900,38 @@ function show_not_finite_fields(Q::MPIStateArray)
...
+    end
+end
+function checked_wait(
+    device::CUDADevice,
+    event,
+    progress = nothing,
+    check = false,
+)
+    wait(device, event, progress)
+end
```
Comment (jkozdon) [inline src/Arrays/MPIStateArrays.jl:935]: Where does the GPU exception get caught? (if this works on the GPU?)

**Review by jkozdon — COMMENTED** [review]
Overall this seems reasonable. I haven't had a chance to run the code yet, but how does the GPU error get caught? Also, I assume that when this is used parallel performance takes a big hit? So should that be noted?

**Comment by kpamnany** [comment]
> Overall this seems reasonable. I haven't had a chance to run the code yet, but how does the GPU error get caught? Also, I assume that when this is used parallel performance takes a big hit? So should that be noted?

`wait`s on `CUDADevice` are asynchronous with respect to the host, so they cannot `throw`; any exceptions are gathered and carried along until a `wait` on `CPU` is issued (which for us happens in `end_ghost_exchange`). At this point, those exceptions are thrown to us. So yes, this does work on the GPU.

Also yes, the `Allreduce` will hurt performance though I haven't quantified how much. I'll add help text about this though, good point.

**Pull request merged**